extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devdb/DevDBProcessor.java

```diff
@@ -0,0 +1,225 @@
+package io.quarkus.datasource.deployment.devdb;
+
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.eclipse.microprofile.config.ConfigProvider;
+import org.jboss.logging.Logger;
+
+import io.quarkus.bootstrap.classloading.QuarkusClassLoader;
+import io.quarkus.datasource.deployment.spi.DefaultDataSourceDbKindBuildItem;
+import io.quarkus.datasource.deployment.spi.DevDBProvider;
+import io.quarkus.datasource.deployment.spi.DevDbConfigurationHandlerBuildItem;
+import io.quarkus.datasource.deployment.spi.DevDbResultBuildItem;
+import io.quarkus.datasource.runtime.DataSourceBuildTimeConfig;
+import io.quarkus.datasource.runtime.DataSourcesBuildTimeConfig;
+import io.quarkus.deployment.IsNormal;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.builditem.RunTimeConfigurationDefaultBuildItem;
+import io.quarkus.deployment.builditem.ServiceStartBuildItem;
+import io.quarkus.deployment.pkg.builditem.CurateOutcomeBuildItem;
+import io.quarkus.runtime.LaunchMode;
+
+public class DevDBProcessor {
+
+    private static final Logger log = Logger.getLogger(DevDBProcessor.class);
+
+    static volatile List<Closeable> databases;
+
+    static volatile Map<String, String> cachedProperties;
+
+    static volatile boolean first = true;
+
+    @BuildStep(onlyIfNot = IsNormal.class)
+    DevDbResultBuildItem launchDatabases(CurateOutcomeBuildItem curateOutcomeBuildItem,
+            List<DefaultDataSourceDbKindBuildItem> installedDrivers,
+            List<DevDBProviderBuildItem> devDBProviders,
+            DataSourcesBuildTimeConfig dataSourceBuildTimeConfig,
+            LaunchModeBuildItem launchMode,
+            BuildProducer<RunTimeConfigurationDefaultBuildItem> runTimeConfigurationDefaultBuildItemBuildProducer,
+            List<DevDbConfigurationHandlerBuildItem> configurationHandlerBuildItems,
+            BuildProducer<ServiceStartBuildItem> serviceStartBuildItemBuildProducer) {
+        //figure out if we need to shut down and restart existing databases
+        //if not and the DB's have already started we just return
+        if (databases != null) {
+            boolean restartRequired = launchMode.getLaunchMode() == LaunchMode.TEST;
+            if (!restartRequired) {
```

Comment:
How does this behave when test profiles are used?

Reply:
It will shut down and recreate the database each time.

I am not sure if this is the correct behaviour or not, it is certainly the safest but may slow things down a bit.

Reply:
I think it's the expected behavior